ArduCopter/mode_auto.cpp

```diff
@@ -771,33 +771,99 @@ void Copter::ModeAuto::wp_run()
         set_throttle_takeoff();
         return;
     }
-
-    // process pilot's yaw input
     float target_yaw_rate = 0;
-    if (!copter.failsafe.radio) {
-        // get pilot's desired yaw rate
-        target_yaw_rate = get_pilot_desired_yaw_rate(channel_yaw->get_control_in());
-        if (!is_zero(target_yaw_rate)) {
-            auto_yaw.set_mode(AUTO_YAW_HOLD);
-        }
-    }
 
-    // set motors to full range
+    //Enable course locked add a2sandres
+    if (g.enbl_crs_lock == 1) {
+        if (!copter.failsafe.radio) {
+            // get pilot's desired yaw rate
+            target_yaw_rate = get_pilot_desired_yaw_rate(channel_yaw->get_control_in());
+            if (!is_zero(target_yaw_rate)) {
+                auto_yaw.set_mode(AUTO_YAW_LOOK_AT_NEXT_WP);
+            }
+        }
+    }
+    else {
+        if (!copter.failsafe.radio) {
+            // get pilot's desired yaw rate
+            target_yaw_rate = get_pilot_desired_yaw_rate(channel_yaw->get_control_in());
+            if (!is_zero(target_yaw_rate)) {
+                auto_yaw.set_mode(AUTO_YAW_HOLD);
+            }
+        }
+    }
+
+   // set motors to full range
     motors->set_desired_spool_state(AP_Motors::DESIRED_THROTTLE_UNLIMITED);
 
-    // run waypoint controller
-    copter.failsafe_terrain_set_status(wp_nav->update_wpnav());
+    // run waypoint controller 
+    copter.failsafe_terrain_set_status(wp_nav->update_wpnav());  
+      
+     // mode semi auto
+    if (g.auto_man_alt == 1) {
+        
+        // get pilot desired climb rate alt control Alt_Hold add a2sAndres
+        float target_climb_rate = get_pilot_desired_climb_rate(channel_throttle->get_control_in());
+        target_climb_rate = constrain_float(target_climb_rate, -get_pilot_speed_dn(), g.pilot_speed_up);
+
+        // get pilot desired lean angles Althold add a2sAndres
+        float target_roll, target_pitch;
+        get_pilot_desired_lean_angles(target_roll, target_pitch, copter.aparm.angle_max, attitude_control->get_althold_lean_angle_max());
+
+        //added control to resume flight add a2sAndres
+        if (target_roll != 0 || target_pitch != 0) {
```

Comment:
What is this needed for? We never changed modes...

Reply:
added control to resume the return plan, this may be due to loss of information, once a height change of more or less than 5 meters has been reached, whereby the aircraft stops its advance forward. To resume the flight plan, enter the pitch or roll.

If you know any solution or a better change for this situation, I would appreciate it if you would tell me. but I have to tell you that we have already done tests with fumigational drones and with frames F550 to check this behavior and for the moment it has worked with the expected results, however we continue doing more rigorous tests for other activities and to be able to implement this behavior with more trust